metrics/sender.go

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package metrics
+
+import (
+	"time"
+
+	log "github.com/cihub/seelog"
+)
+
+// RoleProvider indicates that event was sent from service provider
+const RoleProvider = "provider"
+
+const appName = "myst"
+const startupEventName = "startup"
+
+// Sender builds events and sends them using given transport
+type Sender struct {
+	Transport Transport
+}
+
+// Transport allows sending events
+type Transport interface {
+	sendEvent(event) error
+}
+
+type event struct {
+	Application applicationInfo `json:"application"`
+	EventName   string          `json:"eventName"`
+	CreatedAt   int64           `json:"createdAt"`
+	Context     interface{}     `json:"context"`
+}
+
+type applicationInfo struct {
+	Name    string `json:"name"`
+	Version string `json:"version"`
+}
+
+type startupContext struct {
+	Role string `json:"role"`
+}
+
+// SendStartupEvent sends startup event
+func (sender *Sender) SendStartupEvent(role, version string) {
+	appInfo := applicationInfo{Name: appName, Version: version}
+	event := event{Application: appInfo, EventName: startupEventName, Context: startupContext{Role: role}, CreatedAt: time.Now().Unix()}
+
+	sender.sendEvent(event)
+}
+
+func (sender *Sender) sendEvent(event event) {
+	go func() {
```

Comment:
you'd better off getting rid of the goroutine here. Just return the error and let the consumer decide if he wants to send it in a separate routine or not.

Reply:
I don't want to block application initiation just because statistics are being sent - sending statistics should be done in the background by default.

Now it probably doesn't make a lot of sense here, but later on, we'll have to implement retrying mechanism - especially for connection events, which happen when network is changing and often those requests fail.

That's why I am testing that sender doesn't block execution - with retry mechanism, we'll just have to extend those tests, responsibility of managing execution will be kept in sender.

Reply:
Maybe actually having retry mechanism in this PR would solve this? Because I understand that responsibilities distribution in current code is not that clear :D

Reply:
If you get rid of the goroutine the component becomes flexible, it can be used in synchronous execution and as a background task. Here you're hardcoding it as a background task with no say on the consumer side. 

You can just happily use go ...sendEvent(...) and have the behaviour you want. 


As far as retrying goes - it would probably be best to leave it for the consumer to decide. Maybe some metrics are more relevant than others and they should be retried, while others can be allowed to fail once in a while? My point here is that the consumer should be the one making these assumptions, not the underlying sender.

Reply:
Yeah, flexibility is important and I would like to keep requests being async by default.

Maybe we can have both - what if we had `AsyncTransport` just for making requests in the background?

Both `AsyncTransport(transport Transport)` and `ElasticSearchTransport(url)` would implement same `Transport` interface, so they could be composed together.

By default, we could construct `Sender` with `AsyncTransport(ElasticSearchTransport)` to get async functionality by default, but we can change that later if we need to do sync events somewhere.

Later we can extend `AsyncTransport` to `RetryTransport` (or whatever, I am bad at naming :D) - nothing else would change, just this transport.

Would you be fine with this approach?

Reply:
This misses the point though, you'll have two separate dependencies depending on the behaviour you want. I'd rather have a single component that handles the metrics sending in the way I as the consumer would like. Maybe I want a single request to be sent asynchronously and another one sent synchronously. 

I'm not gonna block the PR for this though.

Reply:
Rule of thumb in go is not to hide go routine in a function, not to make assumptions on how function works. Is a matter of caller how he wants to run it.
As of additional transport interfaces - I don't think there is a point in increasing complexity if not strictly required atm.